applications/StructuralMechanicsApplication/custom_response_functions/adjoint_elements/adjoint_finite_difference_cr_beam_element_3D2N.cpp

```diff
@@ -0,0 +1,287 @@
+        rResult[index]     = geom[i].GetDof(ADJOINT_DISPLACEMENT_X).EquationId();
+        rResult[index + 1] = geom[i].GetDof(ADJOINT_DISPLACEMENT_Y).EquationId();
+        rResult[index + 2] = geom[i].GetDof(ADJOINT_DISPLACEMENT_Z).EquationId();
+
+        rResult[index + 3] = geom[i].GetDof(ADJOINT_ROTATION_X).EquationId();
+        rResult[index + 4] = geom[i].GetDof(ADJOINT_ROTATION_Y).EquationId();
+        rResult[index + 5] = geom[i].GetDof(ADJOINT_ROTATION_Z).EquationId();
+    }
+
+}
+
+void AdjointFiniteDifferenceCrBeamElement::GetDofList(DofsVectorType& rElementalDofList,
+    ProcessInfo& rCurrentProcessInfo) {
+
+    GeometryType & geom = this->GetGeometry();
+
+    const int number_of_nodes = geom.PointsNumber();
+    const int dimension = geom.WorkingSpaceDimension();
+    const unsigned int local_size = number_of_nodes * dimension * 2;
+
+    if (rElementalDofList.size() != local_size) {
+        rElementalDofList.resize(local_size);
+    }
+
+    for (int i = 0; i < number_of_nodes; ++i)
+    {
+        int index = i * number_of_nodes * dimension;
+        rElementalDofList[index]     = geom[i].pGetDof(ADJOINT_DISPLACEMENT_X);
+        rElementalDofList[index + 1] = geom[i].pGetDof(ADJOINT_DISPLACEMENT_Y);
+        rElementalDofList[index + 2] = geom[i].pGetDof(ADJOINT_DISPLACEMENT_Z);
+
+        rElementalDofList[index + 3] = geom[i].pGetDof(ADJOINT_ROTATION_X);
+        rElementalDofList[index + 4] = geom[i].pGetDof(ADJOINT_ROTATION_Y);
+        rElementalDofList[index + 5] = geom[i].pGetDof(ADJOINT_ROTATION_Z);
+    }
+}
+
+void AdjointFiniteDifferenceCrBeamElement::GetValuesVector(Vector& rValues, int Step)
+{
+    KRATOS_TRY
+    const GeometryType & geom = this->GetGeometry();
+    const int number_of_nodes = geom.PointsNumber();
+    const int dimension = geom.WorkingSpaceDimension();
+    const unsigned int num_dofs = number_of_nodes * dimension * 2;
+
+    if (rValues.size() != num_dofs)
+        rValues.resize(num_dofs, false);
+
+    for (int i = 0; i < number_of_nodes; ++i)
+    {
+        const SizeType index = i * dimension * 2;
+
+        const array_1d<double,3>& disp = geom[i].FastGetSolutionStepValue(ADJOINT_DISPLACEMENT, Step);
+        const array_1d<double,3>& rot = geom[i].FastGetSolutionStepValue(ADJOINT_ROTATION, Step);
+
+        rValues[index]     = disp[0];
+        rValues[index + 1] = disp[1];
+        rValues[index + 2] = disp[2];
+
+        rValues[index + 3] = rot[0];
+        rValues[index + 4] = rot[1];
+        rValues[index + 5] = rot[2];
+    }
+    KRATOS_CATCH("")
+}
+
+void AdjointFiniteDifferenceCrBeamElement::Calculate(const Variable<Vector >& rVariable,
+                        Vector& rOutput,
+                        const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY;
+
+    if(rVariable == STRESS_ON_GP || rVariable == STRESS_ON_NODE)
+    {
+        TracedStressType traced_stress_type = static_cast<TracedStressType>(this->GetValue(TRACED_STRESS_TYPE));
```

Comment:
where are those declared?

Reply:
They are declared in the response_data.h.